pink/barriers/barrier.py

```diff
@@ -0,0 +1,221 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+#
+# SPDX-License-Identifier: Apache-2.0
+# Copyright 2022 Stéphane Caron
+
+"""All barriers derive from the :class:`Barrier` base class.
+
+The formalism used in this implementation is written down in
+examples/barriers/NOTES.md
+"""
+
+import abc
+from typing import Callable, Optional, Tuple, Union
+
+import numpy as np
+
+from ..configuration import Configuration
+
+
+class Barrier(abc.ABC):
+    r"""Abstract base class for barrier.
+
+    A barrier is a function :math:`\boldsymbol{h}(\boldsymbol{q})` that
+    satisfies the following condition:
+
+    .. math::
+
+        \frac{\partial \boldsymbol{h}_j}{\partial \boldsymbol{q}}
+        \dot{\boldsymbol{q}} +\alpha_j(\boldsymbol{h}_j(\boldsymbol{q}))
+        \geq 0, \quad \forall j
+
+    where :math:`\frac{\partial \boldsymbol{h}_j}{\partial \boldsymbol{q}}`
+    are the Jacobians of the constraint functions, :math:`\dot{\boldsymbol{q}}`
+    is the joint velocity vector, and :math:`\alpha_j` are extended
+    class K functions.
+
+    Attributes:
+        dim: Dimension of the barrier.
+        gain: barrier gain.
+        class_k_fn: Extended class K function.
+        safe_policy: Safe backup control policy.
+        r: Weighting factor for the safe backup policy regularization term.
+    """
+
+    gain: np.ndarray
+    class_k_fns: Callable[[float], float]
+    safe_policy: np.ndarray
+    r: float
+
+    def __init__(
+        self,
+        dim: int,
+        gain: Union[float, np.ndarray] = 1.0,
+        class_k_fn: Optional[Callable[[float], float]] = None,
+        r: float = 3.0,
+    ):
+        """Initialize the barrier.
+
+        Args:
+            dim: Dimension of the barrier.
+            gain: barrier gain. Defaults to 1.0.
+            class_k_fn: Extended class K function.
+                Defaults to the identity function.
+            r: Weighting factor for the safe backup policy regularization term.
```

Comment:
If I understand well, `r` is a cost regularization weight. Can we make this parameter name into a full word then, for instance `regularization` or `safe_policy_weight` (to emphasize the link with ``self.safe_policy``)?

Reply:
Actually, regarding the expression "safe policy" :grin: I suggest calling it ``safe_velocity`` or ``safe_displacement`` for consistency with the rest of Pink (because we don't use the word "policy" elsewhere). (Also displacement = velocity * dt; we mostly use displacements in Pink.) What do you think?